File: back-end/src/main/java/es/ull/project/application/usecase/serviceassignment/UpdateServiceAssignmentUseCase.java
package es.ull.project.application.usecase.serviceassignment;

import java.util.UUID;
import es.ull.project.domain.entity.ServiceAssignment;
import es.ull.project.domain.valueobject.cost.TransportationVariableCost;
import es.ull.project.domain.valueobject.demand.WasteDemand;
import es.ull.project.domain.valueobject.location.Distance;
import es.ull.project.domain.valueobject.location.ServiceTime;

/**
 * Use case: update of a {@link ServiceAssignment}.
 */
public interface UpdateServiceAssignmentUseCase {

    /**
     * Updates an existing assignment.
     *
     * @param id               identifier of the assignment to update
     * @param containerId      identifier of the container
     * @param facilityId       identifier of the facility
     * @param newWasteDemand   new waste demand
     * @param newDistance      new calculated distance
     * @param newServiceTime   new service time
     * @param newTransportCost new transport cost
     * @return the updated assignment
     */
    ServiceAssignment update(UUID id, UUID containerId, UUID facilityId, WasteDemand newWasteDemand, Distance newDistance, ServiceTime newServiceTime, TransportationVariableCost newTransportCost);
}

File: front-end/src/application/usecase/ServiceAssignmentManagement/UpdateServiceAssignmentUseCase.ts
// UpdateServiceAssignmentUseCase.ts
// Use case contract for updating a service assignment

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { ServiceAssignment } from '../../../domain/entity/ServiceAssignment';
import { WasteDemand } from '../../../domain/valueobject/demand/WasteDemand';
import { Distance } from '../../../domain/valueobject/location/Distance';
import { ServiceTime } from '../../../domain/valueobject/location/ServiceTime';
import { TransportationVariableCost } from '../../../domain/valueobject/cost/TransportationVariableCost';

/**
 * Use case for updating an existing service assignment.
 * Input: assignmentId and fields to update
 * Output: ServiceAssignment (entity)
 */

// Command object for input data
export interface UpdateServiceAssignmentCommand {
    assignmentId: string;
    updatedFields: Partial<{
        wasteDemand: WasteDemand;
        distance: Distance;
        serviceTime: ServiceTime;
        transportCost: TransportationVariableCost;
    }>;
}

// Result type for the use case
export type UpdateServiceAssignmentResult = ServiceAssignment;

// Use case contract
export interface UpdateServiceAssignmentUseCase {
    /**
     * Handles the update of an existing service assignment
     *
     * @param command Data needed to update the service assignment
     * @returns Either a DataError or the updated ServiceAssignment
     */
    execute(command: UpdateServiceAssignmentCommand): Promise<Either<DataError, UpdateServiceAssignmentResult>>;
}
